unstructured_ingest/v2/processes/connectors/vectara.py

```diff
@@ -0,0 +1,318 @@
+            "customer-id": self.connection_config.customer_id,
+            "X-source": "unstructured",
+        }
+
+        response = requests.request(
+            method=http_method, url=url, headers=headers, params=params, data=json.dumps(data)
+        )
+        response.raise_for_status()
+        return response.json()
+
+    def _delete_doc(self, doc_id: str) -> None:
+        """
+        Delete a document from the Vectara corpus.
+
+        Args:
+            url (str): URL of the page to delete.
+            doc_id (str): ID of the document to delete.
+        """
+        body = {
+            "customer_id": self.connection_config.customer_id,
+            "corpus_id": self.connection_config.corpus_id,
+            "document_id": doc_id,
+        }
+        self._request(endpoint="delete-doc", data=body)
+
+    def _index_document(self, document: t.Dict[str, t.Any]) -> None:
+        """
+        Index a document (by uploading it to the Vectara corpus) from the document dictionary
+        """
+        body = {
+            "customer_id": self.connection_config.customer_id,
+            "corpus_id": self.connection_config.corpus_id,
+            "document": document,
+        }
+
+        try:
+            result = self._request(endpoint="index", data=body, http_method="POST")
+        except Exception as e:
+            logger.info(f"exception {e} while indexing document {document['documentId']}")
+            return
+
+        if (
+            "status" in result
+            and result["status"]
+            and (
+                "ALREADY_EXISTS" in result["status"]["code"]
+                or (
+                    "CONFLICT" in result["status"]["code"]
+                    and "Indexing doesn't support updating documents"
+                    in result["status"]["statusDetail"]
+                )
+            )
+        ):
+            logger.info(f"document {document['documentId']} already exists, re-indexing")
+            self._delete_doc(document["documentId"])
+            result = self._request(endpoint="index", data=body, http_method="POST")
+            return
+
+        if "status" in result and result["status"] and "OK" in result["status"]["code"]:
+            logger.info(f"indexing document {document['documentId']} succeeded")
+        else:
+            logger.info(f"indexing document {document['documentId']} failed, response = {result}")
+
+    def write_dict(self, *args, docs_list: t.List[t.Dict[str, t.Any]], **kwargs) -> None:
+        logger.info(f"inserting / updating {len(docs_list)} documents to Vectara ")
+        for vdoc in docs_list:
+            self._index_document(vdoc)
+
+    # def write(self, docs: t.List[BaseIngestDoc]) -> None:
+    def run(
+        self,
+        path: Path,
+        file_data: FileData,
+        **kwargs: t.Any,
+    ) -> Path:
+        docs_list: t.Dict[t.Dict[str, t.Any]] = []
+
+        def get_metadata(element) -> t.Dict[str, t.Any]:
```

Comment:
It looks like there's a lot of data manipulation happening in the uploader, yet this is the intent for the stager.